### DISCUSSION TITLE
About Local Density Penalty Described in the Original RePlace Paper

### DISCUSSION THREAD

**Discussion by by1e11** [The-OpenROAD-Project/OpenROAD#3562] [Q&A] [answered]
Hi, guys

After reading the paper and browsing the source codes, I cannot figure out which parts of the codes reflect the local density penalty described in the paper, could you please suggest where can I find that in the implementation?

Thanks a lot for the nice work.

Bo

**Comment by maliberty** [comment]
@mgwoo can you answer this?

**Comment by mangohehe** [comment]
A method related density computation seems to be removed in the CL https://github.com/The-OpenROAD-Project/OpenROAD/commit/0f82f83049733938c2fb5d6141b291409aa4678f , in which the original design calls updateDensityForceBin inside NesterovPlace::doNesterovPlace. (not sure if this can be related)

@ahmadelrouby could you help understand the changes? thanks

**Reply by mangohehe** [reply]
sorry, never mind, false alarm. 
the function now lives in nesterovbase.

**Comment by mgwoo** [comment]
I didn't decipher the local density features because of unstable issues. 
Please check the standalone repo for the local density calculation:  https://github.com/The-OpenROAD-Project/RePlAce/blob/7c07cffdf419c4dbd315c73c235ec5d670fb0240/src/ns.cpp#L1011

**Reply by mangohehe** [reply]
@mgwoo sorry, i guess I am missing some context.

sounds like "local density penalty" won't be applied due to some unstable issues. I am curious about the issues, what're they, any history documented anywhere?

i am new to the codebase, is that recommended to learn the placer using that standalone repo, which seems achieved without updates for years?  I guess there isn't improvement on the "local density penalty" in Openroad/gpl, isn't it? https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/gpl/src

**Reply by mgwoo** [reply, marked as answer]
When I started the process of deciphering the standalone repo and rewriting the clean C++ source codes from the encrypted variable names and messy structures, one of the RePlAce authors told me that there might be some unstable issues if I bring dynamic step size adaption (flag -DS) and local density penalty (flag -LD) features. 

He rather suggests just focusing on rewriting the pure ePlace Nesterov engine as a placer's starting point. As a result, ePlace engine is available in (OpenROAD/src/gpl), but dynamic step and local density penalty are still missing.

Both local density penalty and dynamic step size adaption (from the RePlAce paper) are not deciphered and are not available in OpenROAD/src/gpl (so, not tested). Those two features are only available in the legacy standalone repo. 

If anyone could decipher those codes and bring them into OpenROAD's RePlAce, it'd be great. However, I'm also not so sure about the ROI. The current OpenROAD's RePlAce is more robust and more stable compared with the standalone, so unstable issues may disappear.

**Reply by mangohehe** [reply]
Thanks for sharing the insights.

**Reply by by1e11** [reply]
That makes great sense. Through our experiences, current RePlace implementation is stable in most cases. An exception could be when a block has low utilization and great density variation.
I will give the local density a try and see if we can get a better placement.

Thanks for the explanation.
